packages/blockchain-wallet-v4-frontend/src/data/components/fundRecovery/sagas.ts

```diff
@@ -0,0 +1,178 @@
+import * as Bitcoin from 'bitcoinjs-lib'
+// eslint-disable-next-line
+import Task from 'data.task'
+import { call, put, select } from 'redux-saga/effects'
+
+import { crypto } from 'blockchain-wallet-v4/src'
+import { selectAll } from 'blockchain-wallet-v4/src/coinSelection'
+import * as Coin from 'blockchain-wallet-v4/src/coinSelection/coin'
+import { APIType } from 'blockchain-wallet-v4/src/network/api'
+import { UnspentResponseType } from 'blockchain-wallet-v4/src/network/api/btc/types'
+import { signSelection as bchSign } from 'blockchain-wallet-v4/src/signer/bch'
+import { HDAccountList, Wallet } from 'blockchain-wallet-v4/src/types'
+import { errorHandler } from 'blockchain-wallet-v4/src/utils'
+import { fromCashAddr, toCashAddr } from 'blockchain-wallet-v4/src/utils/bch'
+import { actions, selectors } from 'data'
+import { promptForSecondPassword } from 'services/sagas'
+
+import * as A from './actions'
+
+export default ({ api }: { api: APIType }) => {
+  // const logLocation = 'components/fundRecovery/sagas'
+  const network = Bitcoin.networks.bitcoin
+  const taskToPromise = (t) => new Promise((resolve, reject) => t.fork(reject, resolve))
+
+  const recoverFunds = function* (action: ReturnType<typeof A.recoverFunds>) {
+    const { payload } = action
+    const {
+      accountIndex,
+      badChange,
+      coin,
+      fromDerivationType,
+      recoveryAddress,
+      unspent_outputs
+    } = payload
+    try {
+      yield put(A.recoverFundsLoading(coin))
+      const password = yield call(promptForSecondPassword)
+
+      if (coin === 'BCH') {
+        const wallet = selectors.core.wallet.getWallet(yield select())
+        const accounts = Wallet.selectHDAccounts(wallet)
+        const account = HDAccountList.selectAccount(accountIndex, accounts)
+        const fromDerivation = account.derivations.find((d) => d.type === fromDerivationType)
+
+        const xprivT = Wallet.isDoubleEncrypted(wallet)
+          ? crypto.decryptSecPass(
+              Wallet.selectSharedKey(wallet),
+              Wallet.selectIterations(wallet),
+              password,
+              fromDerivation.xpriv
+            )
+          : Task.of(fromDerivation.xpriv)
+
+        const xpriv = yield call(() => taskToPromise(xprivT))
+
+        const fee = yield call(api.getBchFees)
+        const coins = unspent_outputs.map((val) => {
+          const path = val.xpub
```

Comment:
@mpfluger for BCH the change goes to address derived from segwit xpub on correct change_index so we need to look up by address, so we don't have the xpub.